README.md
# node-nmea

[![npm version](https://img.shields.io/npm/v/@drivetech/node-nmea.svg)](https://www.npmjs.com/package/@drivetech/node-nmea)
[![npm downloads](https://img.shields.io/npm/dm/@drivetech/node-nmea.svg)](https://www.npmjs.com/package/@drivetech/node-nmea)
[![Build Status](https://travis-ci.org/Drivetech/node-nmea.svg?branch=master)](https://travis-ci.org/Drivetech/node-nmea)
[![Coverage Status](https://coveralls.io/repos/github/Drivetech/node-nmea/badge.svg?branch=master)](https://coveralls.io/github/Drivetech/node-nmea?branch=master)
[![Maintainability](https://api.codeclimate.com/v1/badges/0dd7acaf25e6b4a6d2e3/maintainability)](https://codeclimate.com/github/Drivetech/node-nmea/maintainability)
[![dependencies Status](https://david-dm.org/drivetech/node-nmea/status.svg)](https://david-dm.org/drivetech/node-nmea)
[![devDependencies Status](https://david-dm.org/drivetech/node-nmea/dev-status.svg)](https://david-dm.org/drivetech/node-nmea?type=dev)

> Parser for NMEA sentences.

Available sentences:
* GPRMC - recommended minimum data for gps
* GPGGA - time, position, and fix related data

Example: `$GPRMC,161006.425,A,7855.6020,S,13843.8900,E,154.89,84.62,110715,173.1,W,A*30`

Where:

Value         | Definition
--------------| ----------
161006.425    | Time in UTC 16:10:06.425
A             | Status A=active or V=Void.
7855.6020,S   | Latitude 78°55.6020' N
13843.8900,E  | Longitude 138°43.8900' E
154.89        | Speed over the ground in knots
84.62         | Track angle in degrees True
110715        | Date - 11 of July 2015
173.1,W       | Magnetic Variation in degrees (- West Declination, + East Declination)
A             | FAA Mode A=autonomous, D=differential, E=estimated (dead-reckoning), M=manual input, S=simulated, N=data not valid, P=precise (4.00 and later)
\*30          | The checksum data, always begins with \*

Example: `$GPGGA,120558.916,5058.7457,N,00647.0514,E,2,06,1.7,109.0,M,47.6,M,1.5,0000*71`

Where:

Value         | Definition
--------------| ----------
120558.916    | Time in UTC 12:05:58.916 
5058.7457,N   | Latitude 50°58.7457' N
00647.0514,E  | Longitude 6°47.0514' E
2             | Gps quality. 0=Invalid, 1=GPS fix (SPS), 2=DGPS fix, 3=PPS fix, 4=Real Time Kinematic, 5=Float RTK, 6=estimated (dead reckoning) (2.3 feature), 7=Manual input mode, 8=Simulation mode
06            | Number of satellites
1.7           | HDOP
109.0,M       | Altitude in meters
47.6,M        | Geoidal Separation in meters
1.5           | Age Gps data in seconds
0000          | Reference Station Id
\*71          | The checksum data, always begins with \*

## Installation

```bash
$ npm install @drivetech/node-nmea
```

## Use

[Try on RunKit](https://npm.runkit.com/@drivetech/node-nmea)
```js
const nmea = require('@drivetech/node-nmea')

const raw = '$GPRMC,161006.425,A,7855.6020,S,13843.8900,E,154.89,84.62,110715,173.1,W,A*30'
const data = nmea.parse(raw)
data.valid // true
data.raw // '$GPRMC,161006.425,A,7855.6020,S,13843.8900,E,154.89,84.62,110715,173.1,W,A*30'
data.type // 'RMC'
data.gps // true
data.datetime // Sat Jul 11 2015 13:10:06 GMT-0300 (CLT)
data.loc // { geojson: { type: 'Point', coordinates: [ 138.7315, -78.9267 ] }, dmm: { latitude: '7855.6020,S', longitude: '13843.8900,E' } }
data.speed // { knots: 154.89, kmh: 286.85627999999997 }
data.track // '84.62'
data.magneticVariation // '173.1,W'
data.mode // 'Autonomous'
```

## License

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
## [2.0.1](https://github.com/drivetech/node-nmea/compare/v2.0.0...v2.0.1) (2019-01-02)


### Bug Fixes

* **package:** transfer to drivetech org ([37336ae](https://github.com/drivetech/node-nmea/commit/37336ae))

# [2.0.0](https://github.com/lgaticaq/node-nmea/compare/v1.0.4...v2.0.0) (2018-10-17)


### Features

* **node:** update min version to v8 ([8ab5575](https://github.com/lgaticaq/node-nmea/commit/8ab5575))


### BREAKING CHANGES

* **node:** set min node version to v8

#### 1.0.4 (2017-06-26)

##### Chores

* **package:**
  * update dependencies ([326acf3b](https://github.com/lgaticaq/node-nmea/commit/326acf3b3f68eb0ea452279f7f7d46b7752fb71f))
  * update eslint to version 4.1.0 ([704c960a](https://github.com/lgaticaq/node-nmea/commit/704c960a80b284344159d11f0d2c89d5f74b9cbf))
  * update chai to version 4.0.1 ([63356d20](https://github.com/lgaticaq/node-nmea/commit/63356d20e43132db116495c70a307f83c3adf0f8))

#### 1.0.3 (2016-10-21)

##### Chores

* **test:** Add more test ([f0d5a82b](https://github.com/lgaticaq/node-nmea/commit/f0d5a82b5ba519f8f0ffb35c4f1b85a11510d116))
* **package:**
  * Add yarn ([9ddd271b](https://github.com/lgaticaq/node-nmea/commit/9ddd271bea4f6ab121b6fc46618120fc15f56898))
  * Update dependencies ([bd9cfc87](https://github.com/lgaticaq/node-nmea/commit/bd9cfc871fb1edfe562dccc667c2ecb0b6949447))
  * update codeclimate-test-reporter to version 0.4.0 ([2409818c](https://github.com/lgaticaq/node-nmea/commit/2409818caf3f8ede7e279540531369658102a77e))

##### Bug Fixes

* **pad:** Fix parse latToDmm and lngToDmm ([92f33a2a](https://github.com/lgaticaq/node-nmea/commit/92f33a2a3ea434f4f8a5bf5162119fb1671ed188))
* **parser:** Parse to float track data ([0e530450](https://github.com/lgaticaq/node-nmea/commit/0e530450947b74652ca127e5cb0dd484bc6b2b26))

#### 1.0.2 (2016-8-11)

##### Chores

* **package:**
  * Remove pad ([11e94c03](https://github.com/lgaticaq/node-nmea/commit/11e94c03a37baadb5a66efb0053ac0fd12d7e440))
  * Update dependencies ([05fb3329](https://github.com/lgaticaq/node-nmea/commit/05fb3329d259de8169018eae70ad3c87b5cc6314))
  * update mocha to version 3.0.0 ([e7a30540](https://github.com/lgaticaq/node-nmea/commit/e7a30540b49e6cca7786799d3b7f400005510644))
  * update eslint to version 3.0.0 ([53af3297](https://github.com/lgaticaq/node-nmea/commit/53af32976784fc66ef24246876d134eaa3fccba6))
* **src:**
  * Remove moment ([4f6b991c](https://github.com/lgaticaq/node-nmea/commit/4f6b991ca9bb850874acfd0a7edf10b43813f3bb))
  * Remove moment ([175203d2](https://github.com/lgaticaq/node-nmea/commit/175203d2bb55c1aa59be2acb234565ac57c77936))

#### 1.0.1 (2016-6-24)

##### Bug Fixes

* **parser:** Fix get deg, min and sign in degToDec ([f2caa0d2](https://github.com/lgaticaq/node-nmea/commit/f2caa0d2d427749211d447907cb3e32006db5f74))

## 1.0.0 (2016-6-21)

##### Chores

* **node:** Remove babel and change min node version to v4 ([68f8b0b6](https://github.com/lgaticaq/node-nmea/commit/68f8b0b6b5d24639d37734b6c28af4047ee66f01))
* **package:**
  * update mocha to version 2.5.0 ([feb15c6e](https://github.com/lgaticaq/node-nmea/commit/feb15c6e54ab10f02b8d0d2e5e0e92c2b3534847))
  * update babel-core to version 6.7.7 ([f06d7561](https://github.com/lgaticaq/node-nmea/commit/f06d7561ef4a55d18163687992391dfe78ef282b))
  * update babel-cli to version 6.8.0 ([9bca3ce4](https://github.com/lgaticaq/node-nmea/commit/9bca3ce41611dcf159d2ae686ae9c9442feb607a))

##### Bug Fixes

* **parser:** Update regex to variations of RMC ([5956101f](https://github.com/lgaticaq/node-nmea/commit/5956101fb67ff14551c86eab209fa03641457ae1))

#### 0.6.2 (2016-03-20)

* Add dependencies badge ([4a87497](https://github.com/lgaticaq/node-nmea/commit/4a87497))
* Add gpgga in keywords ([c070123](https://github.com/lgaticaq/node-nmea/commit/c070123))
* Remove extra describes ([ddbce35](https://github.com/lgaticaq/node-nmea/commit/ddbce35))
* Update dependencies ([8b919b5](https://github.com/lgaticaq/node-nmea/commit/8b919b5))
* Update node versions ([3bcd4fd](https://github.com/lgaticaq/node-nmea/commit/3bcd4fd))

#### 0.6.1 (2016-02-22)

* Fix track regex ([66bd80f](https://github.com/lgaticaq/node-nmea/commit/66bd80f))

#### 0.6.0 (2016-02-12)

* Add support to GPGGA ([8b7ca94](https://github.com/lgaticaq/node-nmea/commit/8b7ca94))

#### 0.5.0 (2016-02-11)

* Add lat and lng in dmm ([1c5e7c1](https://github.com/lgaticaq/node-nmea/commit/1c5e7c1))
* Fix readme ([f000c31](https://github.com/lgaticaq/node-nmea/commit/f000c31))
* Fix test ([e618eba](https://github.com/lgaticaq/node-nmea/commit/e618eba))
* Remove unused dependencies ([ea3394f](https://github.com/lgaticaq/node-nmea/commit/ea3394f))

#### 0.4.0 (2016-02-11)

* Add description for magnetic variation ([50b16e1](https://github.com/lgaticaq/node-nmea/commit/50b16e1))
* Add extra to npmignore ([f97aa24](https://github.com/lgaticaq/node-nmea/commit/f97aa24))
* Add node version ([890bcc6](https://github.com/lgaticaq/node-nmea/commit/890bcc6))
* Add tonic example ([5f7b01b](https://github.com/lgaticaq/node-nmea/commit/5f7b01b))
* Fix readme ([e5408e2](https://github.com/lgaticaq/node-nmea/commit/e5408e2))
* Remove eslint config ([1425099](https://github.com/lgaticaq/node-nmea/commit/1425099))
* Remove random ([b6626de](https://github.com/lgaticaq/node-nmea/commit/b6626de))
* Update dependencies ([042688f](https://github.com/lgaticaq/node-nmea/commit/042688f))
* Update test ([4b6f673](https://github.com/lgaticaq/node-nmea/commit/4b6f673))

#### 0.3.1 (2015-09-03)

* Fix latToDmm, lngToDmm and degToDec. ([3e95ed9](https://github.com/lgaticaq/node-nmea/commit/3e95ed9))

#### 0.3.0 (2015-08-28)

* Fix parse whit gps status void. Add gps to parse data. Fix #6 ([48aea87](https://github.com/lgaticaq/node-nmea/commit/48aea87)), closes [#6](https://github.com/lgaticaq/node-nmea/issues/6)

#### 0.2.4 (2015-08-27)

* Add node 0.10 to travis ([35a2988](https://github.com/lgaticaq/node-nmea/commit/35a2988))

test/nmea.js
"use strict"

import * as nmea from "../src"
import assert from "assert"
import moment from "moment"

describe("Nmea", () => {
  const data = nmea.randomData()
  const parser = nmea.parse(data.raw)

  describe("#this.raw", () => {
    it("should return the same value passed in the constructor", () => {
      assert.equal(data.raw, parser.raw)
    })
  })

  describe("#this.isValid()", () => {
    it("should return true if checksum is valid", () => {
      assert.equal(true, nmea.isValid(parser.raw))
    })
  })

  describe("#this.datetime", () => {
    it("should return true if datetime values is valid", () => {
      const datetime = moment.utc(parser.datetime)
      const date = datetime.format("DDMMYY")
      assert.equal(data.date, date)
      const time = datetime.format("HHmmss.SSS")
      assert.equal(data.time, time)
    })
  })

  describe("#this.loc", () => {
    it("should return true if position values is valid", () => {
      assert.equal("Point", parser.loc.type)
      assert.equal(2, parser.loc.coordinates.length)
      const lng = nmea.lngToDmm(parser.loc.coordinates[0])
      const lat = nmea.latToDmm(parser.loc.coordinates[1])
      assert.equal(data.latitude, lat)
      assert.equal(data.longitude, lng)
    })
  })
})


test/test.js
'use strict'

const { describe, it } = require('mocha')
const nmea = require('../src')
const expect = require('chai').expect

describe('Nmea', () => {
  describe('GPRMC', () => {
    const data =
      '$GPRMC,194329.000,A,3321.6735,S,07030.7640,W,0.00,0.00,090216,,,A*6C'
    const parser = nmea.parse(data)

    it('should return the same value passed in the constructor', () => {
      expect(data).to.eql(parser.raw)
    })

    it('should return true if checksum is valid', () => {
      expect(nmea.isValid(parser.raw)).to.eql(true)
    })

    it('should return true if position values is valid', () => {
      expect(parser.loc.geojson.type).to.eql('Point')
      expect(parser.loc.geojson.coordinates.length).to.eql(2)
      expect(parser.loc.dmm.latitude).to.eql('3321.6735,S')
      expect(parser.loc.dmm.longitude).to.eql('07030.7640,W')
    })

    it('should return true if gps value is valid', () => {
      expect(parser.gps).to.eql(true)
    })

    it('should return true if speed value is valid', () => {
      expect(parser.speed.knots).to.eql(0)
      expect(parser.speed.kmh).to.eql(0)
    })

    it('should return true if track value is valid', () => {
      expect(parser.track).to.eql(0.0)
    })

    it('should return true if magnetic variation value is valid', () => {
      expect(parser.magneticVariation).to.eql(null)
    })

    it('should return true if mode value is valid', () => {
      expect(parser.mode).to.eql('Autonomous')
    })
  })

  describe('GPRMC alternative', () => {
    const data =
      '$GPRMC,192053.00,A,4212.63658,N,00844.30075,W,0.648,295.88,010616,,,A*76'
    const parser = nmea.parse(data)

    it('should return the same value passed in the constructor', () => {
      expect(data).to.eql(parser.raw)
    })

    it('should return true if checksum is valid', () => {
      expect(nmea.isValid(parser.raw)).to.eql(true)
    })

    it('should return true if position values is valid', () => {
      expect(parser.loc.geojson.type).to.eql('Point')
      expect(parser.loc.geojson.coordinates.length).to.eql(2)
      expect(parser.loc.dmm.latitude).to.eql('4212.63658,N')
      expect(parser.loc.dmm.longitude).to.eql('00844.30075,W')
    })

    it('should return true if gps value is valid', () => {
      expect(parser.gps).to.eql(true)
    })

    it('should return true if speed value is valid', () => {
      expect(parser.speed.knots).to.eql(0.648)
      expect(parser.speed.kmh).to.eql(1.200096)
    })

    it('should return true if track value is valid', () => {
      expect(parser.track).to.eql(295.88)
    })

    it('should return true if magnetic variation value is valid', () => {
      expect(parser.magneticVariation).to.eql(null)
    })

    it('should return true if mode value is valid', () => {
      expect(parser.mode).to.eql('Autonomous')
    })

    it('should return data GPRMC with empty fields', () => {
      const data =
        '$GPRMC,194329.000,A,3321.6735,S,07030.7640,W,,,090216,173.1,W,*50'
      const parser = nmea.parse(data)
      expect(parser.speed.knots).to.eql(null)
      expect(parser.speed.kmh).to.eql(null)
      expect(parser.track).to.eql(null)
      expect(parser.magneticVariation).to.eql('173.1,W')
      expect(parser.mode).to.eql(null)
    })
  })

  describe('GPGGA', () => {
    const data =
      '$GPGGA,172814.0,3723.46587704,N,12202.26957864,W,2,6,1.2,18.893,M,-25.669,M,2.0,0031*4F'
    const parser = nmea.parse(data)

    it('should return the same value passed in the constructor', () => {
      expect(data).to.eql(parser.raw)
    })

    it('should return true if checksum is valid', () => {
      expect(nmea.verifyChecksum(parser.raw)).to.eql(true)
    })

    it('should return true if position value is valid', () => {
      expect(parser.loc.geojson.type).to.eql('Point')
      expect(parser.loc.geojson.coordinates.length).to.eql(2)
      expect(parser.loc.dmm.latitude).to.eql('3723.46587704,N')
      expect(parser.loc.dmm.longitude).to.eql('12202.26957864,W')
    })

    it('should return true if gps quality value is valid', () => {
      expect(parser.gpsQuality).to.eql('DGPS fix')
    })

    it('should return true if satellites value is valid', () => {
      expect(parser.satellites).to.eql(6)
    })

    it('should return true if hdop value is valid', () => {
      expect(parser.hdop).to.eql(1.2)
    })

    it('should return true if altitude values is valid', () => {
      expect(parser.altitude).to.eql(18.893)
    })

    it('should return true if geoidal separation values is valid', () => {
      expect(parser.geoidalSeparation).to.eql(-25.669)
    })

    it('should return true if age Gps Data values is valid', () => {
      expect(parser.ageGpsData).to.eql(2)
    })

    it('should return true if reference station id values is valid', () => {
      expect(parser.refStationId).to.eql('0031')
    })

    it('should return data GPGGA with empty fields', () => {
      const data =
        '$GPGGA,172814.0,3723.46587704,N,12202.26957864,W,2,6,,,,,,,0031*4F'
      const parser = nmea.parse(data)
      expect(parser.hdop).to.eql(null)
      expect(parser.altitude).to.eql(null)
      expect(parser.geoidalSeparation).to.eql(null)
      expect(parser.ageGpsData).to.eql(null)
    })
  })

  describe('Extras', () => {
    it('should return latitude in dmm', () => {
      expect(nmea.latToDmm(-33.38113666666667)).to.eql('3322.8682,S')
      expect(nmea.lngToDmm(-70.77628166666666)).to.eql('07046.5769,W')
      expect(nmea.latToDmm(-33)).to.eql('3300.0000,S')
      expect(nmea.lngToDmm(-70)).to.eql('07000.0000,W')
      expect(nmea.latToDmm(33)).to.eql('3300.0000,N')
      expect(nmea.lngToDmm(70)).to.eql('07000.0000,E')
      expect(nmea.kmhToKnots()).to.eql(0.0)
      expect(nmea.kmhToKnots(100)).to.eql(53.99568034557235)
    })
  })

  describe('Error', () => {
    it('should return invalid data', () => {
      const data =
        'GGA,172814.0,3723.46587704,N,12202.26957864,W,2,6,1.2,18.893,M,-25.669,M,2.0,0031'
      const parser = nmea.parse(data)
      expect(parser.raw).to.eql(data)
      expect(parser.valid).to.eql(false)
    })
    it('should return invalid data', () => {
      const data =
        '$GPRMC,194329.000,A,3321.6735,S,07030.7640,W,0.00,0.00,090216,,,A*6F'
      const parser = nmea.parse(data)
      expect(parser.raw).to.eql(data)
      expect(parser.valid).to.eql(false)
    })
  })
})


src/test/nmea.js
"use strict"

import nmea from "../"
import assert from "assert"
import moment from "moment"

describe("Nmea", () => {
  const data = nmea.randomData()
  const parser = nmea.parse(data.raw)

  describe("#this.raw", () => {
    it("should return the same value passed in the constructor", () => {
      assert.equal(data.raw, parser.raw)
    })
  })

  describe("#this.isValid()", () => {
    it("should return true if checksum is valid", () => {
      assert.equal(true, nmea.isValid(parser.raw))
    })
  })

  describe("#this.datetime", () => {
    it("should return true if datetime values is valid", () => {
      const datetime = moment.utc(parser.datetime)
      const date = datetime.format("DDMMYY")
      assert.equal(data.date, date)
      const time = datetime.format("HHmmss.SSS")
      assert.equal(data.time, time)
    })
  })

  describe("#this.loc", () => {
    it("should return true if position values is valid", () => {
      assert.equal("Point", parser.loc.type)
      assert.equal(2, parser.loc.coordinates.length)
      const lng = nmea.lngToDmm(parser.loc.coordinates[0])
      const lat = nmea.latToDmm(parser.loc.coordinates[1])
      assert.equal(data.latitude, lat)
      assert.equal(data.longitude, lng)
    })
  })
})


src/index.js
'use strict'

const pad = (n, width, z) => {
  n = n + ''
  return n.length >= width ? n : new Array(width - n.length + 1).join(z) + n
}

/**
 * Get checksum from raw data
 *
 * @param {string} data - raw data
 * @return {string} checksum en hex
 */
const getChecksum = data => {
  let checksum
  let idx1 = data.indexOf('$GP')
    if (idx1 === -1){
      idx1 = data.indexOf('$GN')
    }
  const idx2 = data.indexOf('*')
  checksum = data
    .slice(idx1 + 1, idx2)
    .split('')
    .reduce((y, x) => y ^ x.charCodeAt(0), 0)
  return checksum
}

/**
 * Verify checksum from raw data
 *
 * @param {string} data - raw data
 * @return {boolean} if valid data
 */
const verifyChecksum = data => {
  const idx = data.indexOf('*')
  return getChecksum(data) === parseInt(data.substr(idx + 1, 2), 16)
}

/**
 * regex for GPRMC valid data
 */
const gprmc = /^\$G[NP](\w{3}),(\d{6}([.]\d+)?),([AV]),(\d{4}([.]\d+)?,[NS]),(\d{5}([.]\d+)?,[WE]),(\d{1,3}[.]\d{1,3})?,(\d{1,3}[.]\d{1,3})?,(\d{6}),((\d{1,3}[.]\d{1,3})?,([WE])?),?([ADENS])?\*([0-9A-F]{2})$/
/**
 * regex for GPGGA valid data
 */
const gpgga = /^\$G[NP](\w{3}),(\d{6}([.]\d+)?),(\d{4}[.]\d+,[NS]),(\d{5}[.]\d+,[WE]),([0-8]),(\d{1,2}),(\d{1,3}[.]\d{1,3})?,([-]?\d+([.]\d+)?)?,M?,([-]?\d+([.]\d+)?)?,M?,(\d+([.]\d+)?)?,(\d{4})?,?([ADENS])?\*([0-9A-F]{2})$/

const gngll = /^\$G[NP](\w{3}),(\d{4}[.]\d+,[NS]),(\d{5}[.]\d+,[WE]),(\d{6}([.]\d+)?),([AV]),?([ADENS])?\*([0-9A-F]{2})$/

/**
 * Verify if raw data is valid
 *
 * @param {string} data - raw data
 * @return {boolean} if valid data
 */
const isValid = data => {
 return gprmc.test(data) && verifyChecksum(data)
}

/**
 * Decimal latitude to degree [dmm]
 *
 * @param {string} data - raw data
 * @return {string} degree [dmm]
 */
const latToDmm = data => {
  const tmp = data.toString().split('.')
  const deg = pad(Math.abs(tmp[0]), 2, '0')
  const mim = pad((('0.' + (tmp[1] || 0)) * 60).toFixed(4), 7, '0')
  const sign = data < 0 ? 'S' : 'N'
  return `${deg}${mim},${sign}`
}

/**
 * Decimal longitude to degree [dmm]
 *
 * @param {string} data - raw data
 * @return {string} degree [dmm]
 */
const lngToDmm = data => {
  const tmp = data.toString().split('.')
  const deg = pad(Math.abs(tmp[0]), 3, '0')
  const mim = pad((('0.' + (tmp[1] || 0)) * 60).toFixed(4), 7, '0')
  const sign = data < 0 ? 'W' : 'E'
  return `${deg}${mim},${sign}`
}

/**
 * Degree [dmm] to decimal
 *
 * @param {string} data - Degree in dmm.
 * @return {number} decimals
 */
const degToDec = data => {
  let decimal = 0.0
  const _data = data.match(/(\d{2,3})(\d{2}[.]\d+),([NSWE])/).slice(1)
  const deg = _data[0]
  const min = _data[1]
  const sign = _data[2]
  decimal = parseFloat(deg) + parseFloat(min) / 60
  if (sign === 'S' || sign === 'W') {
    decimal *= -1
  }
  return decimal
}

/**
 * Knots to Km/h
 *
 * @param {string} data - knots
 * @return {number} km/h
 */
const knotsToKmh = knots => parseFloat(knots) * 1.852

/**
 * km/h to knots
 *
 * @param {number} data - km/h
 * @return {number} knots
 */
const kmhToKnots = kmh => {
  let knots = 0.0
  if (kmh) {
    knots = kmh / 1.852
  }
  return knots
}

/**
 * FAA modes
 *
 */
const faaModes = {
  A: 'Autonomous',
  D: 'Differential',
  E: 'Estimated',
  M: 'Manual input',
  S: 'Simulated',
  N: 'Not Valid',
  P: 'Precise'
}

const gpsQualities = {
  '0': 'Invalid',
  '1': 'GPS fix (SPS)',
  '2': 'DGPS fix',
  '3': 'PPS fix',
  '4': 'Real Time Kinematic',
  '5': 'Float RTK',
  '6': 'estimated (dead reckoning) (2.3 feature)',
  '7': 'Manual input mode',
  '8': 'Simulation mode'
}

/**
 * Parse GPRMC raw data
 *
 * @param {string} raw - raw data
 * @return {object} data parse
 */
const parseRmc = raw => {
  let data = { raw: raw, valid: false }
  const r = gprmc.exec(raw)
  if (isValid(raw)) {
    data.type = r[1]
    const datetime = `${r[11]}${r[2]}`
    const pattern = /(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})[.]\d{1,3}/
    data.datetime = new Date(
      datetime.replace(pattern, '20$3-$2-$1T$4:$5:$6.000Z')
    )
    data.loc = {
      geojson: {
        type: 'Point',
        coordinates: [degToDec(r[7]), degToDec(r[5])]
      },
      dmm: {
        latitude: r[5],
        longitude: r[7]
      }
    }
    data.gps = r[4] === 'A'
    data.speed = {
      knots: r[9] ? parseFloat(r[9]) : null,
      kmh: r[9] ? knotsToKmh(r[9]) : null
    }
    data.track = r[10] ? parseFloat(r[10]) : null
    data.magneticVariation = r[12] === ',' ? null : r[12]
    data.mode = r[15] ? faaModes[r[15]] : null
    data.valid = true
  }
  return data
}

/**
 * Parse GPGGA raw data
 *
 * @param {string} raw - raw data
 * @return {object} data parse
 */
const parseGga = raw => {
  let data = { raw: raw, valid: false }
  const r = gpgga.exec(raw)
  data.raw = raw
  data.type = r[1]
  const now = new Date()
  const date = now.toISOString().split('T')[0]
  const pattern = /(\d{2})(\d{2})(\d{2})[.](\d{1,3})/
  data.datetime = new Date(`${date}T${r[2].replace(pattern, '$1:$2:$3')}.000Z`)
  data.loc = {
    geojson: {
      type: 'Point',
      coordinates: [degToDec(r[5]), degToDec(r[4])]
    },
    dmm: {
      latitude: r[4],
      longitude: r[5]
    }
  }
  data.gpsQuality = gpsQualities[r[6]]
  data.satellites = parseInt(r[7], 10)
  data.hdop = r[8] ? parseFloat(r[8]) : null
  data.altitude = r[9] ? parseFloat(r[9]) : null
  data.geoidalSeparation = r[11] ? parseFloat(r[11]) : null
  data.ageGpsData = r[13] ? parseFloat(r[13]) : null
  data.refStationId = r[15]
  data.valid = verifyChecksum(r[0])
  return data
}

/**
 * Parse GNGLL raw data
 *
 * @param {string} raw - raw data
 * @return {object} data parse
 */
const parseGll = raw => {
  let data = { raw: raw, valid: false }
  const r = gngll.exec(raw)
  data.raw = raw
  data.type = r[1]
  const now = new Date()
  const date = now.toISOString().split('T')[0]
  const pattern = /(\d{2})(\d{2})(\d{2})[.](\d{1,3})/
  data.datetime = new Date(`${date}T${r[4].replace(pattern, '$1:$2:$3')}.000Z`)
  data.loc = {
    geojson: {
      type: 'Point',
      coordinates: [degToDec(r[3]), degToDec(r[2])]
    },
    dmm: {
      latitude: r[2],
      longitude: r[3]
    }
  }
  data.valid = verifyChecksum(r[0])
  return data
}


const parse = raw => {
  let data = { raw: raw, valid: false }
  if (gprmc.test(raw)) {
    data = parseRmc(raw)
  } else if (gpgga.test(raw)) {
    data = parseGga(raw)
  }
    else if (gngll.test(raw)){
    data = parseGll(raw)
    }
  return data
}

module.exports = {
  getChecksum: getChecksum,
  verifyChecksum: verifyChecksum,
  gprmc: gprmc,
  isValid: isValid,
  latToDmm: latToDmm,
  lngToDmm: lngToDmm,
  degToDec: degToDec,
  knotsToKmh: knotsToKmh,
  kmhToKnots: kmhToKnots,
  faaModes: faaModes,
  parseRmc: parseRmc,
  parseGga: parseGga,
  parse: parse
}